### ISSUE TITLE
Chunksize FutureWarning cannot be silenced without affecting "chunks"

### ISSUE THREAD

**Issue by kmuehlbauer** [h5netcdf/h5netcdf#147]
#### What happened:

While fixing h5netcdf-changes over in xarray https://github.com/pydata/xarray/pull/6200 I found that those FutureWarnings cannot be silenced keeping current behavior.

- `chunks=None` -> `chunks=None` with `FutureWarning` 
- `chunks="h5py"` -> `chunks=True`
- `chunks="h5netcdf"` -> `chunks=_get_default_chunksizes`

#### What you expected to happen:

Silencing the `FutureWarning` keeping current behavior should be possible.

@dionhaefner Could you take a look here?

**Comment by dionhaefner** [comment]
Actually I'm not quite sure whether we want to do `chunks="h5py"` -> `chunks=None`. The way I see it the proper migration is this:

```python
f.dimensions = {"x": 1, "time": None}

# OLD

# uses h5py chunking and throws warning
f.create_dataset("foo", ("x", "time"))  # could also be chunks=True

# no unlimited dims -> no warning
f.create_dataset("foo2", ("x",))  # could also be chunks=True

# NEW

# keep using h5py chunks, no warning
f.create_dataset("foo", ("x", "time"), chunks="h5py")

# no unlimited dims -> no warning (unchanged)
f.create_dataset("foo2", ("x",))
```

The distinction between `chunks="h5py"` and `chunks="h5netcdf"` is only relevant if unlimited dimensions are present. In this case, chunking is mandatory anyway. So the question boils down to: what do we want this to do?

```python
# no unlimited dimensions present, so this would be unchunked by default
f.create_dataset("foo2", ("x",), chunks="h5py")
```

I think the path of least confusion is to treat this like `chunks=True`. Since something is passed to the `chunks` keyword I would interpret this as "the user wants chunking". But if you disagree I'm fine with changing it to h5py's `chunks=None`.

**Comment by kmuehlbauer** [comment]
@dionhaefner I appreciate this comprehensive representation of the different possibilities. I mostly agree and I think your outlined solution might be the way to go. But do you have a solution at hand for those users, who do not want to use the new default chunking after we switch to it?

**Comment by dionhaefner** [comment]
That's essentially my last code snippet, which boils down to

1. If your variable has unlimited dimensions, use `chunks=h5py`.
2. If not, just keep using what you are using now.

This recipe should work in all cases to keep the current behavior.

For the record, I also like your suggestion of a `chunking_heuristic` keyword that would side-step these subtleties, and that could go through a similar depreciation cycle (start with `chunking_heuristic=None` which defaults to `h5py` and emits a warning if unlimited dimensions are present, then switch to `chunking_heuristic="h5netcdf"` at 1.0).

**Comment by kmuehlbauer** [comment]
Thanks! Let's reconvene early next week for a decision.

**Comment by kmuehlbauer** [comment]
I the light of upcoming `h5netcdf v1.0` our solution should be bullet proof to be sustainable over the lifetime of `v1.0`.

As we've seen handling those subtleties can be quite challenging and it might lead into a dead-end at some point in time. There is another solution which should at least be considered, an options-object. Several Python packages are using config/options one way or the other and in our situation it would really help simplifying things.

So my suggestion would still be using `chunk_heuristics`, but as a kwarg of that options-object.

### Minimal implementation:

```python
H5NETCDF_OPTIONS = {"chunk_heuristics": "h5py"}

class set_options:
    """Set options for h5netcdf.

    Parameters
    ----------
    "chunk_heuristics : {"h5py", "h5netcdf"}, default: "h5py"
        Heuristics used, if chunking is needed or requested.
    """

    def __init__(self, **kwargs):
        self.last_options = {}
        for k, v in kwargs.items():
            if k not in H5NETCDF_OPTIONS:
                raise ValueError(
                    f"{k!r} is not a valid option. Valid options are {set(H5NETCDF_OPTIONS)}."
                )
            chunk_heuristics_options = set(["h5py", "h5netcdf"])
            if k == "chunk_heuristics" and v not in chunk_heuristics_options:
                errmsg = f"Expected one of {chunk_heuristics_options!r}"
            else:
                errmsg = ""
            if "Expected" in errmsg:
                raise ValueError(f"invalid value {v!r} for option {k!r}. " + errmsg)
            self.last_options[k] = H5NETCDF_OPTIONS[k]
        self._update_options(kwargs)

    def _update_options(self, new_options):
        H5NETCDF_OPTIONS.update(new_options)

    def __enter__(self):
        return

    def __exit__(self, type, value, traceback):
        self._update_options(self.last_options)


def get_options():
    """Get options for h5netcdf."""
    return Frozen("H5NETCDF_OPTIONS")
```

### use current default ("h5py"), will trigger the future warning

```python
import h5netcdf
with h5netcdf.File(tmp_local_netcdf, "w") as ds:
    ds.dimensions = {"x": 10, "y": 10, "z": 10, "t": None}
    v = ds.create_variable(
        "hello2", ("x", "y", "z", "t"), "float", chunks=None
    )
```

### set options "h5py" (will not trigger the future warning)

```python
import h5netcdf
h5netcdf.set_options(chunk_heuristics="h5py)
with h5netcdf.File(tmp_local_netcdf, "w") as ds:
    ds.dimensions = {"x": 10, "y": 10, "z": 10, "t": None}
    v = ds.create_variable(
        "hello2", ("x", "y", "z", "t"), "float", chunks=None
    )
```

### use "h5py" with context (will not trigger the future warning)

```python
import h5netcdf
with h5netcdf.set_options(chunk_heuristics="h5py):
  with h5netcdf.File(tmp_local_netcdf, "w") as ds:
      ds.dimensions = {"x": 10, "y": 10, "z": 10, "t": None}
      v = ds.create_variable(
          "hello2", ("x", "y", "z", "t"), "float", chunks=None
      )
```

The above will work the same way with `chunk_heuristics="h5netcdf"`.

The major advantage over our other possible solutions is that almost no code has to be changed. It boils down to only one line -> setting the option.

**Comment by dionhaefner** [comment]
I think global options make sense (and I use them in a lot of projects), but IMO it's a bit awkward if you only have 1 option to set :)

If you have several things in mind that would make sense to access through it, sure, go for it. Otherwise I think it's still good design to use the `chunk_heuristic` keyword, and add that to a global options object later if there is demand for one. I don't think people usually have dozens of different calls to `create_variable` with unlimited dimensions, so it's probably a O(1) line change anyway.

**Comment by dionhaefner** [comment]
And in terms of bullet proofness, I agree. So we probably *should* implement that `chunk_heuristic` keyword, and not overload `chunks`. But whether or not that can also be set through an options object is a separate discussion, isn't it?

**Comment by kmuehlbauer** [comment]
> And in terms of bullet proofness, I agree. So we probably _should_ implement that `chunk_heuristic` keyword, and not overload `chunks`.

Yes, it would be very difficult handling an overloaded `chunks` keyword.

> But whether or not that can also be set through an options object is a separate discussion, isn't it?

Yes, we can separate that and move that in at any later point.

**Comment by kmuehlbauer** [comment]
Sorry, fat fingers...

**Comment by dionhaefner** [comment]
Do you want me to do that in #150 or will you do it in #148?

**Comment by kmuehlbauer** [comment]
> Do you want me to do that in #150 or will you do it in #148?

I can do it tomorrow in #148. If you have additional cycles to do it in #150, please go ahead. Thanks.

**Comment by kmuehlbauer** [comment]
Resolved with #150
